File: py/train.py
import matplotlib.pyplot as plt
import os

import torch
from torch import nn
from torch.utils.data import DataLoader

from dataset import MNISTDataset
from model import NeuralNetwork, train, test

CACHE_DIR = "./.cache"

batch_size = 32 

train_dataset = MNISTDataset("./data/train-labels.idx1-ubyte", "./data/train-images.idx3-ubyte")
train_dataloader = DataLoader(train_dataset, batch_size=batch_size)

test_dataset = MNISTDataset("./data/t10k-labels.idx1-ubyte", "./data/t10k-images.idx3-ubyte")
test_dataloader = DataLoader(test_dataset, batch_size=batch_size)

for X, y in test_dataloader:
    print(f"Shape of X [N, C, H, W]: {X.shape}")
    print(f"Shape of y: {y.shape} {y.dtype}")
    break

# Display image and label.
train_features, train_labels = next(iter(train_dataloader))
print(f"Feature batch shape: {train_features.size()}")
print(f"Labels batch shape: {train_labels.size()}")
# img = train_features[0].squeeze()
# label = train_labels[0]
# plt.imshow(img, cmap="gray")
# plt.show()
# print(f"Label: {label}")

device = torch.accelerator.current_accelerator().type if torch.accelerator.is_available() else "cpu"
print(f"Using {device} device")

model = NeuralNetwork().to(device)
print(model)

loss_fn = nn.CrossEntropyLoss()
optimizer = torch.optim.SGD(model.parameters(), lr=1e-3)

epochs = 50
for t in range(epochs):
    print(f"Epoch {t+1}\n-------------------------------")
    train(train_dataloader, model, loss_fn, optimizer, device)
    test(test_dataloader, model, loss_fn, device)
print("Done!")

os.makedirs(CACHE_DIR, exist_ok=True)
torch.save(model.state_dict(), CACHE_DIR + "/model.pth")
print("Saved PyTorch Model State to model.pth")

File: src/train.cpp
#include "./utils/include/idx_parser.h"
#include "./utils/include/model.h"
#include "./utils/include/model_parser.h"
#include <iostream>
#include <utility>

int main() {
  const std::string CACHE_DIR = "./.cache";
  const std::string DATA_DIR = "./data";
  IdxParser images_parser(DATA_DIR + "/train-images.idx3-ubyte");
  IdxParser labels_parser(DATA_DIR + "/train-labels.idx1-ubyte");
  ModelParser model_parser(CACHE_DIR + "/params_cpp.bin");

  NeuralNetwork network(model_parser.dims);
  // for (auto dim : model_parser.dims)
  //   std::cout << dim.first << '\n';
  network.load_params(&model_parser);

  // std::vector<std::pair<int, int>> dims = {std::make_pair(512, 28 * 28), std::make_pair(512, 512), std::make_pair(10, 512)};
  // NeuralNetwork network(dims);
  // network.initialize_params();

  int batch_size = 64;
  int num_batches = 10;
  int num_epochs = 50;
  // std::vector<float> images(images_parser.element_size * batch_size * num_batches);
  // std::vector<uint8_t> labels(labels_parser.element_size * batch_size * num_batches);

  std::vector<float> images(images_parser.element_size * images_parser.length);
  std::vector<uint8_t> labels(labels_parser.element_size * labels_parser.length);
  images_parser.read(images, images.size());
  labels_parser.read(labels, labels.size());

  network.gradient_descent(images, labels, batch_size, num_epochs);
  network.save_params(CACHE_DIR + "/params_cpp.bin");
}
